torch/csrc/autograd/function.h

```diff
@@ -49,6 +51,53 @@ struct FunctionFlags {
   function_list next_functions;
 };
 
+namespace {
+
+inline void variable_function_flags(FunctionFlags& f, size_t i, const Variable& var) {
+  if (!var.defined()) return;
+  f.is_executable |= var.requires_grad();
+  if (var.grad_fn()) {
+    f.next_functions[i] = std::make_pair<>(var.grad_fn(), var.output_nr());
+  } else if (var.requires_grad()) {
+    f.next_functions[i] = std::make_pair<>(var.grad_accumulator(), 0);
+  }
+}
+
+template<typename... Args> inline void set_variable_function_flags(FunctionFlags& flags, size_t i);
+template<typename... Args> inline void set_variable_function_flags(FunctionFlags& flags, size_t i, const Variable& x, Args... args);
+template<typename... Args> inline void set_variable_function_flags(FunctionFlags& flags, size_t i, at::ArrayRef<Variable> xs, Args... args);
+
+template<typename... Args> inline void set_variable_function_flags(FunctionFlags& flags, size_t i) {}
+template<typename... Args> inline void set_variable_function_flags(FunctionFlags& flags, size_t i, const Variable& x, Args... args) {
+  variable_function_flags(flags, i++, x);
+  set_variable_function_flags(flags, i, args...);
+}
+template<typename... Args> inline void set_variable_function_flags(FunctionFlags& flags, size_t i, at::ArrayRef<Variable> xs, Args... args) {
+  for (const auto& x : xs) {
+    variable_function_flags(flags, i++, x);
+  }
+  set_variable_function_flags(flags, i, args...);
+}
+
+// Sigh... such duplication.
+template<typename... Args> inline void set_tensor_function_flags(FunctionFlags& flags, size_t i);
+template<typename... Args> inline void set_tensor_function_flags(FunctionFlags& flags, size_t i, const at::Tensor& x, Args... args);
+template<typename... Args> inline void set_tensor_function_flags(FunctionFlags& flags, size_t i, at::ArrayRef<at::Tensor> xs, Args... args);
```

Comment:
Why do you have to pre-declare all overloads?

Reply:
I suppose you technically do not need all of the predeclarations; you only need to predeclare the last function: https://godbolt.org/g/FhkLPs (and in general, you need to predeclare all declarations after the first function which makes a recursive call).